ensemble_md/utils/coordinate_swap.py

```diff
@@ -0,0 +1,1493 @@
+    Rx = [[1, 0, 0, 0], [0, c / d, -b / d, 0], [0, b / d, c / d, 0], [0, 0, 0, 1]]
+    Rx_r = [[1, 0, 0, 0], [0, c / d, b / d, 0], [0, -b / d, c / d, 0], [0, 0, 0, 1]]
+
+    # Rotation about the y axis
+    Ry = [[d, 0, -a, 0], [0, 1, 0, 0], [a, 0, d, 0], [0, 0, 0, 1]]
+    Ry_r = [[d, 0, a, 0], [0, 1, 0, 0], [-a, 0, d, 0], [0, 0, 0, 1]]
+
+    # Rotation about the z axis
+    Rz = [[np.cos(angle), np.sin(angle), 0, 0], [-np.sin(angle), np.cos(angle), 0, 0], [0, 0, 1, 0], [0, 0, 0, 1]]
+
+    # Initial position
+    x0, y0, z0 = point
+    init_arry = [x0, y0, z0, 1]
+
+    # Perform the transofmration
+    final_arry = T_r@Rx_r@Ry_r@Rz@Ry@Rx@T@init_arry
+
+    # Get the corrdinates for the rotated point
+    rotated_point = final_arry[0:3]
+
+    return rotated_point
+
+
+def find_rotation_angle(initial_point, vertex, rotated_point, axis):
+    """
+    Determines the angle of rotation around an arbitrary axis in 3D space.
+
+    Parameters
+    ----------
+    initial_point: numpy.ndarray
+        The initial point coordinates.
+    rotated_point: numpy.ndarray
+        The rotated point coordinates.
+    axis: A numpy array
+        The axis of rotation.
+
+    Returns
+    -------
+    angle: int
+        The angle of rotation in radians.
+    """
+    # Normalize axis and seperate coordinates
+    u, v, w = axis / np.linalg.norm(axis)
+    x0, y0, z0 = initial_point
+    a, b, c = vertex
+
+    # Determine the rotated point with the same bond length as the initial point
+    rotated_point_axis = rotated_point - vertex
+    norm_rot_point_vec = rotated_point_axis / np.linalg.norm(rotated_point_axis)
+    final_bond_length = np.sqrt((x0 - a) ** 2 + (y0 - b) ** 2 + (z0 - c) ** 2)
+    rotated_point_standard = vertex + final_bond_length*norm_rot_point_vec
+    x, y, z = rotated_point_standard  # Change to point with same bond length along this vector
+
+    Q = np.sqrt(v ** 2 + w ** 2)
+    B = ((w / Q) * (y0 - b)) - ((v / Q) * (z0 - c))
+    A = Q * (x0 - a) - u * (((v / Q) * (y0 - b)) + ((w / Q) * (z0 - c)))
+    C = u * (x0 - a) + Q * (((v / Q) * (y0 - b)) + ((w / Q) * (z0 - c)))
+    L = np.arctan2(B, A)
+
+    # Determine the rotational angle to line-up point
+    cos_arg = (x - a - u * C) / (Q * np.sqrt(A ** 2 + B ** 2))
+    if cos_arg > 1 or cos_arg < -1:
+        if abs(cos_arg - 1) < .0001:
+            cos_arg = 1
+        elif abs(cos_arg + 1) < .0001:
+            cos_arg = -1
+        else:
+            raise Exception(f'Invalid arccos argument: {(x - a - u * C) / (Q * np.sqrt(A ** 2 + B ** 2))}')
+    arc_cos = np.arccos(cos_arg)
+    angle = L + arc_cos
+
+    # Test point
+    new_pt = rotate_point_around_axis(initial_point, vertex, axis / np.linalg.norm(axis), angle)
+    if abs(new_pt[1] - y) < 0.0001:
+        return angle
+    else:
+        arc_cos = 2*np.pi - arc_cos
+        angle = L + arc_cos
+        if angle > 2*np.pi:
+            angle = angle - 2*np.pi
+
+    return angle
+
+
+def compute_angle(coords):

```

Comment:
@ajfriedman22 this function is actually calculating the cosine of the angle instead of the angle in radians indicated in the docstring. Were you expecting angles or cosine values? If you were expecting angle values, it would be worth checking if this would influence your simulations. It seems that `get_missing_coord` is the only function that uses `compute_angle`.

Reply:
So if I understand correctly, a part of the function `get_missing_coord` tries to rotate `init_coords` with different angles until the angle difference between `iangle` and `internal_angle` is smaller than 0.05. (And the cutoff `min_dev_angles` starts from 3.) Assuming that you were expecting $\theta_1-\theta_2 \lt 0.05$ rad but were actually using $\cos\theta_1-\cos\theta_2 \lt 0.05$, the relevant question now is the range of $\theta_1-\theta_1$.

Reply:
So I take the arc cos in line 958 so this should be computing an angle in radians

Reply:
Ugh okay. I did not look into other functions. So you were indeed considering the cutoff based on the difference in the cosine values and it was just that the docstring of `compute_angle` was wrong? If so, I'll update the docstring of `compute_angle`.

Reply:
And "angle" is then added to an arctan output so it should be an angle, but I could be missing something here.

Reply:
I mean what is returned by `compute_angle` is indeed a cosine value instead of an angle in radian, so the docstring of `compute_angle` for the argument `angle`, i.e., `Angle in radians between the two points.` seems wrong. If that is the case, then I'll update the docstring. :)

Reply:
So the arccos is taken within the `compute_angle` function so the docstring should be correct. The intent is for `compute_angle` to return an angle in radians and then later we require that $\theta_1$ - $\theta_2$ < 0.05 as the cut-off for alignment with the existing internal angles of the molecule.

Reply:
Sorry for the confusion here. I was looking at the wrong function. It is resolved now and the error was not present in the separate function used to generate the data for the paper so it must have just been a version issue when I added the code to ensemble_md.

Reply:
No worries! I was just making sure if it was a bug that we should worry about or not.